### PR TITLE
appMode override

### LINKED ISSUE #948: allow app-mode override
Let the caller override the inferred type of content.

Separated from #942.

### PR DESCRIPTION
**Pull request by aronatkins**
Allow override of the inferred `appMode` for situations where complicated deployments can lead to incorrectly inferred content type. For example, the presence of an `app.R` forces the content to be deployed as a Shiny application with the `appMode = "shiny"` even if that file was an auxiliary resource for some other type of content that happens to have lower precedence.

The appMode argument is accepted by:
- `deployApp`
- `writeManifest`
- `listDeploymentFiles`
- `appDependencies`

The `appDependencies` function supports obtaining a set of deployment files through `appFileManifest`.

The `appMetadata` function no longer calls `listDeploymentFiles`. It assumes that the caller has already fully enumerated the required files and provided them in `appFiles`.

Fixes #948

### CODE REVIEW

File: R/deployApp.R [hunk 1]
```diff
@@ -75,8 +75,12 @@
 #'   You can find the `appId` in the following places:
 #'   * On shinyapps.io, it's the `id` listed on the applications page.
 #'   * For Posit Connect, it's `guid` from the info tab on the content page.
-#' @param contentCategory Optional; the kind of content being deployed (e.g.
-#'   `"plot"` or `"site"`).
+#' @param appMode Optional; the type of content being deployed (e.g. `"shiny"`).
```
Comment (kevinushey) [inline R/deployApp.R:78]: Do we need to document all of the different valid `appMode` values?

**Review by kevinushey — APPROVED** [review]

**Review by toph-allen — APPROVED** [review]
Looks good to me.

File: R/deployApp.R (hunk 1 above)
Comment (aronatkins) [reply R/deployApp.R:78]: At this point, I was thinking of this as "expert use only". The accepted values are available in the Connect API reference. rsconnect currently generates a subset of these values. Maybe I'll add something like:

> Accepted values include "shiny", "api", "rmd-static", "quarto-static", and "static".  See the `<Posit Connect API Reference>` for a full set of values. Not all servers support all types of content.

Comment (kevinushey) [reply R/deployApp.R:78]: That looks perfect to me.

**Comment by ChaitaC** [comment]
Edge case: My deployment is stuck without any error message: 
Steps:
1. remotes::install_github("rstudio/rsconnect")
2. Create a new directory 
3. Add aap.R and quarto.qmd in the folder 
4. Deploy without specifying any appmode 
`rsconnect::deployApp(appName = "test-app-mode", server = "rsc.radixu.com", account = "ChaitaTest")`
```
> rsconnect::deployApp(appName = "test-app-mode", server = "rsc.radixu.com", account = "ChaitaTest")
── Preparing for deployment ──────────────────────────────────────────────────
✔ Deploying "test-app-mode" to "server: rsc.radixu.com / username: ChaitaTest"
ℹ Creating application on server...
✔ Created application with id 23056
ℹ Bundling 2 files: app.R and quarto.qmd
ℹ Capturing R dependencies with renv
✔ Found 30 dependencies
✔ Created 16,845b bundle
ℹ Uploading bundle...
```
Incomplete app available here: https://rsc.radixu.com/connect/#/apps/b0ce2571-aef9-4181-81b1-0ebe178cf2e7/access

**Comment by ChaitaC** [comment]
Test ran on ubuntu20-password 
- Deploy app when multiple files in directory without specifying appMode
1. remotes::install_github("rstudio/rsconnect")
2. Create a new directory
3. Add aap.R and quarto.qmd in the folder
4. Deploy without specifying any appmode
`rsconnect::deployApp(appName = "test", server = "192.168.56.2", account = "chaitatest")`
```
rsconnect::deployApp(appName = "test", server = "192.168.56.2", account = "chaitatest")
── Preparing for deployment ──────────────────────────────────────────────────────────────────────────────────────────────
✔ Deploying "test" to "server: 192.168.56.2 / username: chaitatest"
ℹ Creating application on server...
✔ Created application with id 3
ℹ Bundling 2 files: app.R and quarto.qmd
ℹ Capturing R dependencies with renv
✔ Found 30 dependencies
✔ Created 16,847b bundle
ℹ Uploading bundle...
✔ Uploaded bundle with id 4
── Deploying to server ───────────────────────────────────────────────────────────────────────────────────────────────────
Building Shiny application...
```

- Update appMode of existing shinyapp to be deployed as API 
`rsconnect::deployApp(appMode = "api", appName = "test", server = "192.168.56.2", account = "chaitatest")`
```
── Deploying to server ───────────────────────────────────────────────────────────────────────────────────────────────────
Error in `POST()`:
! <http://192.168.56.2:3939/__api__/applications/3/deploy> failed with HTTP status 403
you cannot change the type of content once deployed. This content is 'shiny', but the bundle is identified as 'api'. Try
deploying this bundle as a new content item rather than updating an existing one.
Run `rlang::last_trace()` to see where the error occurred.
```

Deploy as New app
```
> rsconnect::deployApp(appMode = "api", appName = "test123", server = "192.168.56.2", account = "chaitatest")
── Preparing for deployment ──────────────────────────────────────────────────────────────────────────────────────────────
✔ Deploying "test123" to "server: 192.168.56.2 / username: chaitatest"
ℹ Creating application on server...
✔ Created application with id 4
ℹ Bundling 2 files: app.R and quarto.qmd
ℹ Capturing R dependencies with renv
✔ Found 36 dependencies
✔ Created 19,494b bundle
ℹ Uploading bundle...
✔ Uploaded bundle with id 6
── Deploying to server ───────────────────────────────────────────────────────────────────────────────────────────────────
Building Plumber API...
```
Note: Deployment succeeds but app will not work since this is a shiny bundle deployed as API

**Comment by aronatkins** [comment]
IDE work to build upon this change: https://github.com/rstudio/rstudio/pull/13510

**Comment by ChaitaC** [comment]
Testing instructions need update: https://github.com/rstudio/rsconnect/issues/958

**Pull request merged**